synapse/rest/client/login.py

```diff
@@ -342,6 +343,16 @@ async def _complete_login(
             user_id = canonical_uid
 
         device_id = login_submission.get("device_id")
+
+        # Check that device_id is not greater than a reasonable 512B
+        device_id_size = sys.getsizeof(device_id)
+        if device_id_size > 512:
+            raise LoginError(
+                400,
+                "Device_id cannot be greater than 512B.",
```

Comment:
I think the error message could be a point of confusion. On my system, a 464-character ASCII string (which would be encoded as 464 bytes in UTF-8) just breaches the proposed representation size limit:

```python
>>> sys.getsizeof("1"*464)
513
```

So the error message could be a point of confusion here. I'd suggest either:
- be vaguer: say something like "device_id is too long" without committing to specifics
- use a different metric: e.g. reject strings whose `len()`is bigger than 512. That should correspond to strings of at most 512 unicode codepoints (whose raw data would be encoded in UTF-8 in at most 2048 bytes).

Given my slightly paranoid nitpicks in the other comment, I think I'd opt for the latter; but I defer to your and the team's judgement!